protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,235 @@
+
+```pseudo
+team_epoch_volume = sum[min(party_epoch_volume, referralProgram.maxPartyVolumePerEpoch) for each party in team]
+```
+
+After the values are calculated, the `team_epoch_volume` is stored by the network and each parties `party_epoch_volume` is reset to `0` ready for the next epoch.
+
+The network can then calculate the teams `team_running_volume` by summing a teams team_epoch_volume values over the last n epochs where n is the `window_length` set in the [governance proposal](#governance-proposals).
+
+## Benefit mechanics
+
+### Setting benefit factors
+
+Whilst a referral program is `STATUS_ACTIVE`, at the start of an epoch the network must set the `referral_reward_factor` and `referral_discount_factor` for each referee. This is done by identifying a referees team and identifying the teams current benefit tier. A teams benefit tier is defined as the highest tier for which their `team_running_volume` is greater or equal to the tiers `minimum_running_volume`. If a party does not qualify for any tier, both values are set to `0`.
+
+```psuedo
+Given:
+    benefit_tiers=[
+        {
+            "minimum_running_volume": 10000,
+            "referral_reward_factor": 0.001,
+            "referral_discount_factor": 0.001,
+        },
+        {
+            "minimum_running_volume": 20000,
+            "referral_reward_factor": 0.005,
+            "referral_discount_factor": 0.005,
+        },
+        {
+            "minimum_running_volume": 30000,
+            "referral_reward_factor": 0.010,
+            "referral_discount_factor": 0.010,
+        },
+    ]
+
+And:
+    team_running_volume=22353
+
+Then:
+    referral_reward_factor=0.005
+    referral_reward_factor=0.005
+```
+
+These benefit factors are then fixed for the duration of the next epoch.
+
+### Applying benefit factors
+
+Whenever a fee is incurred by a referee (either during continuous trading or on auction exit) the network must apply referral program benefits.
+
+The network can first calculate individual fee components following the [fees specification](./0029-FEES-fees.md#calculating-fees) and apply any usual checks.
+
+The network must then:
+
+- Calculate the rewards due to the referrer.
+    ```pseudo
+    infrastructure_fee_reward = floor(infrastructure_fee * referral_reward_factor) 
+    liquidity_fee_reward = floor(liquidity_fee * referral_reward_factor) 
+    maker_fee_reward = floor(maker_fee * referral_reward_factor) 
+    ```
+- Calculate the discounts due to the referee.
+    ```pseudo
+    infrastructure_fee_discount = floor(infrastructure_fee * referral_discount_factor)
+    liquidity_fee_discount = floor(liquidity_fee * referral_discount_factor)
```

Comment:
I just wonder whether the above are used consistently with the sprit of the numbers in the above example. Because with increasing volume, I see the reward factors decreasing (good) but then I'd expect 

```
floor(liquidity_fee * (1-referral_discount_factor))
```

Reply:
This is the discount no? It would be weird if `discount = x * (1-discount_factor)`

Reply:
Not sure I follow, your equation would give a discounted fee rather the discount to be applied no?

We could do it that way (and then not have to update the fees) but the logic for calculating the discount and then applying it to the fee is so we could then expose the discount value through an API.

(Which I now realise isn't actually in the spec - will add that).

Reply:
Sorry my bad: I was reading it wrong - I was reading new fee (which isn't what it said, sorry), rather than discount. Yes setting discount to this is correct.